internal/unionstore/art/art_iterator.go

```diff
@@ -318,197 +338,166 @@ func (it *baseIter) compare(other *baseIter) int {
-			idx = 0 // mark in-place leaf is visited
-			it.idxes[depth] = idx
-			if !n256.inplaceLeaf.addr.IsNull() {
-				return n256.inplaceLeaf
+		case typeNode256:
+			n256 := it.allocator.getNode256(curr.addr)
+			if idx == inplaceIndex {
+				idx = 0 // mark in-place leaf is visited
+				it.idxes[depth] = idx
+				if !n256.inplaceLeaf.addr.IsNull() {
+					return n256.inplaceLeaf
+				}
+			} else if idx == 256 {
+				break
+			}
+			idx = n256.nextPresentIdx(idx)
+			if idx < 256 {
+				it.idxes[depth] = idx
+				child = &n256.children[idx]
 			}
-		} else if idx == 256 {
-			break
 		}
-		idx = n256.nextPresentIdx(idx)
-		if idx < 256 {
-			it.idxes[depth] = idx
-			child := n256.children[idx]
+		if child != nil {
 			if child.kind == typeLeaf {
 				it.idxes[depth]++
-				return child
+				return *child
 			}
-			it.nodes = append(it.nodes, child)
+			it.nodes = append(it.nodes, *child)
 			it.idxes = append(it.idxes, inplaceIndex)
-			return it.next()
+			continue
 		}
+		it.nodes = it.nodes[:depth]
+		it.idxes = it.idxes[:depth]
+		if depth == 0 {
+			return nullArtNode
+		}
+		it.idxes[depth-1]++
 	}
-	it.nodes = it.nodes[:depth]
-	it.idxes = it.idxes[:depth]
-	if depth == 0 {
-		return nullArtNode
-	}
-	it.idxes[depth-1]++
-	return it.next()
 }
 
 func (it *baseIter) prev() artNode {
-	depth := len(it.nodes) - 1
-	curr := it.nodes[depth]
-	idx := it.idxes[depth]
-	idx--
-	switch curr.kind {
-	case typeNode4:
-		n4 := it.allocator.getNode4(curr.addr)
-		idx = min(idx, int(n4.nodeNum)-1)
-		if idx >= 0 {
-			it.idxes[depth] = idx
-			child := n4.children[idx]
-			if child.kind == typeLeaf {
-				return child
+	for {
+		depth := len(it.nodes) - 1
+		curr := it.nodes[depth]
+		idx := it.idxes[depth]
+		idx--
+		var child *artNode
+		switch curr.kind {
+		case typeNode4:
+			n4 := it.allocator.getNode4(curr.addr)
+			idx = min(idx, int(n4.nodeNum)-1)
+			if idx >= 0 {
+				it.idxes[depth] = idx
+				child = &n4.children[idx]
+			} else if idx == inplaceIndex {
+				it.idxes[depth] = idx
+				if !n4.inplaceLeaf.addr.IsNull() {
+					return n4.inplaceLeaf
+				}
 			}
-			it.nodes = append(it.nodes, child)
-			it.idxes = append(it.idxes, node256cap)
-			return it.prev()
-		} else if idx == inplaceIndex {
-			it.idxes[depth] = idx
-			if !n4.inplaceLeaf.addr.IsNull() {
-				return n4.inplaceLeaf
+		case typeNode16:
+			n16 := it.allocator.getNode16(curr.addr)
+			idx = min(idx, int(n16.nodeNum)-1)
+			if idx >= 0 {
+				it.idxes[depth] = idx
+				child = &n16.children[idx]
+			} else if idx == inplaceIndex {
+				it.idxes[depth] = idx
+				if !n16.inplaceLeaf.addr.IsNull() {
+					return n16.inplaceLeaf
+				}
 			}
-		}
-	case typeNode16:
-		n16 := it.allocator.getNode16(curr.addr)
-		idx = min(idx, int(n16.nodeNum)-1)
-		if idx >= 0 {
-			it.idxes[depth] = idx
-			child := n16.children[idx]
-			if child.kind == typeLeaf {
-				return child
+		case typeNode48:
+			n48 := it.allocator.getNode48(curr.addr)
+			if idx >= 0 {
```

Comment:
Anything needs for the `else` branch of this condition, like `panic`?

Reply:
The else branch means(`idx == inplaceIndex`), it will be handled later. Added some comment.

Reply:
Could the `idx` be other negative values, like in situations the memory is corruted?
```
If 
else if
else
   panic("xx")
```
How about also adding panic for these unrechable branches to make it more clear.